### ISSUE TITLE
linearize-once Gauss-Newton products

### ISSUE THREAD

**Issue by mattjj** [google/jax#529]
Follow-up from the thread in #526.

**Comment by j-towns** [comment]
I guess a 'transpose' operator would be handy for this.

**Comment by mattjj** [comment]
That would do it, though it might also be overkill. Another option, sketched in [this comment](https://github.com/google/jax/issues/526#issuecomment-476424692), would be to return the linearized function from `ad.vjp` (different from `api.vjp`), and then lift it back into Python as in `api.lift_linearized`. In other words, `ad.vjp` already forms the linearized function we want; we just have to return it!

The `ad.backward_pass` function is basically how to transpose a jaxpr. I'm not sure how to expose to the user a `transpose` operator, but I'm sure we can get it with some thought...

**Comment by mattjj** [comment]
There may be an easier way to do this: we could use `linearize` to replace a function with its linear approximation, then compute an HVP against the loss composed with that linear approximation.

**Comment by Ajoo** [comment]
Here's how I've been doing this following your suggestions. I meant to look into it much sooner but life got in the way.

```python
def value_and_grad_and_gvp(fun, loss_fun, argnums=0):
  '''Returns a function that computes the loss, gradient and returns a function
  for computing the Gauss-Newton matrix-vector product at a given point
  '''
  def compute_value_and_grad_and_gvp(*args, **kwargs):
    # make f a function of only the relevant arguments
    f = lu.wrap_init(fun, kwargs)
    f_partial, dyn_args = _argnums_partial(f, argnums, args)

    ans, jvp_fun = linearize(f_partial.call_wrapped, *dyn_args)

    def loss_after_affine_f(*dprimals):
      dans = jvp_fun(*dprimals)
      return loss_fun(ans + dans)

    zero_dprimals = tree_map(np.zeros_like, dyn_args)
    # hvp of loss composed with an affine approximation of fun computed at zero
    g, gvp_fun = linearize(grad(loss_after_affine_f), *zero_dprimals)
    # we could get this computation together with grad and gvp_fun if linearize
    # had a has_aux
    l = loss_fun(ans)

    return l, g, gvp_fun

  return compute_value_and_grad_and_gvp
```

Sorry about the weird api choices. For my application it makes sense to compute the value and grad and, at the same time return a function to compute the gvp w.r.t. the same arguments at the same point.
I also borrowed the `_argnums_partial` function rather than implement another version of it to restrict the vjps and jvps to a particular subset of arguments but it is kind of awkward to use it here together with the `linearize` and `grad` functions... I've also noticed that there are some issues with linearizing (or computing the `jvp`) of a function after `_argnums_partial` in the particular case where the derivatives are 0...

An alternative way to do this would be simply getting the linearization of the function and the `vjp` of this linearization. The gvp can then be computed as usual:

```python
def value_and_grad_and_gvp(fun, loss_fun, argnums=0):
  '''Returns a function that computes the loss, gradient and returns a function
  for computing the Gauss-Newton matrix-vector product at a given point
  '''
  # for some implementation of this function
  val_grad_hvp_loss_fun = value_and_grad_and_hvp(loss_fun)

  def compute_value_and_grad_and_gvp(*args, **kwargs):
    # make f a function of only the relevant arguments
    f = lu.wrap_init(fun, kwargs)
    f_partial, dyn_args = _argnums_partial(f, argnums, args)

    ans, jvp_fun = linearize(f_partial.call_wrapped, *dyn_args)
    # linearization point doesn't matter here as jvp_fun is linear
    # using dyn_args for convenience
    vjp_fun = vjp(jvp_fun, *dyn_args)[1]

    l, grad_loss, hvp_loss_fun = val_grad_hvp_loss_fun(ans)

    #function to compute the gvp
    def gvp_fun(*tangents):
        gv = vjp_fun(hvp_loss_fun(jvp_fun(*tangents)))
        return gv[0] if isinstance(argnums, int) else gv

    g = vjp_fun(grad_loss)
    g = g[0] if isinstance(argnums, int) else g
    return l, g, gvp_fun

  return compute_value_and_grad_and_gvp
```

I guess performance-wise it should be about the same right?

**Comment by Ajoo** [comment]
Btw, I also toyed with the first option you gave. I ended up implementing a jp function that returns both a python function to compute the vjp and the jvp together. (Basically just glued the code for linearize and vjp together)

```python
from jax.api import _check_args, _check_inexact_input_vjp
import jax.linear_util as lu
from jax.util import unzip2, partial
from jax.api_util import pytree_fun_to_jaxtupletree_fun, pytree_to_jaxtupletree, apply_jaxtree_fun
from jax.tree_util import build_tree, PyTreeDef, tree_map
from jax.interpreters import partial_eval as pe
from jax.interpreters import ad
from jax.tree_util import node_types
from jax.core import pack, eval_jaxpr, get_aval, lattice_join, unit

def jp_(traceable, primals, has_aux=False):
  if not has_aux:
    out_primal, pval, jaxpr, consts = ad.linearize(traceable, *primals)
  else:
    out_primal, pval, jaxpr, consts, aux = ad.linearize(
      traceable, *primals, has_aux=True)

  primal_avals = list(map(get_aval, primals))

  def jvp_(*t):
    tangent_avals = list(map(get_aval, t))
    for primal_aval, tangent_aval in zip(primal_avals, tangent_avals):
      try:
        lattice_join(primal_aval, tangent_aval)
      except TypeError:
        msg = ("linearized function called on tangent values inconsistent with "
               "the original primal values.")
        raise ValueError(msg)
    primals = pack(t)  # dummy primals
    tangents = pack(t)
    _, ans = eval_jaxpr(jaxpr, consts, (), primals, tangents)
    return pe.merge_pvals(ans, pval)

  def vjp_(ct):
    ct = ad.ignore_consts(ct, pval)
    dummy_primal_and_ct = pack((unit, ct))
    dummy_args = (None,) * len(jaxpr.invars)
    _, arg_cts = ad.backward_pass(
      jaxpr, consts, (), dummy_args, dummy_primal_and_ct)
    return ad.instantiate_zeros(pack(primals), arg_cts[1])

  if not has_aux:
    return out_primal, jvp_, vjp_
  else:
    return out_primal, jvp_, vjp_, aux


def jp(fun, *primals, has_aux=False):
  '''
  Returns fun evaluated at primals and python functions for evaluating
  both the jvp and vjp
  '''
  if not isinstance(fun, lu.WrappedFun):
    fun = lu.wrap_init(fun)
  primals_flat, in_trees = unzip2(map(pytree_to_jaxtupletree, primals))
  _check_args(primals_flat)
  tree_map(_check_inexact_input_vjp, primals)
  jaxtree_fun, out_tree = pytree_fun_to_jaxtupletree_fun(fun, in_trees)

  if not has_aux:
    out_primal, out_jvp, out_vjp = jp_(jaxtree_fun, primals_flat)
  else:
    out_primal, out_jvp, out_vjp, aux = jp_(
      jaxtree_fun, primals_flat, has_aux=True)
  out_tree = out_tree()
  if has_aux:
    out_tree, aux_tree = out_tree.children
  out_primal_py = build_tree(out_tree, out_primal)
  ct_in_trees = [out_tree]
  ct_out_tree = PyTreeDef(node_types[tuple], None, in_trees)

  jvp_py = partial(apply_jaxtree_fun, out_jvp, (in_trees, out_tree))

  # Not sure why this is in jvp but I'm keeping it just in case...
  def out_vjp_packed(cotangent_in):
    return out_vjp(cotangent_in)
  vjp_py = partial(apply_jaxtree_fun, out_vjp_packed,
              (ct_in_trees, ct_out_tree))

  if not has_aux:
    return out_primal_py, jvp_py, vjp_py
  else:
    return out_primal_py, jvp_py, vjp_py, build_tree(aux_tree, aux)
```

But this approach does seem too complicated if you get the same from computing the vjp of a linearized version of fun as in the second alternative of my previous post.

**Comment by mattjj** [comment]
At one point we added `jax.linear_transpose` for this!
